### DISCUSSION TITLE
non_upstream_failed trigger rule

### DISCUSSION THREAD

**Discussion by felicity-cl** [apache/airflow#22029] [General]
### Description

I'm missing a trigger rule that makes sure a downstream task only runs if all upstream tasks are done, but by done means only  success or failed state.

### Use case/motivation

currently we only have none skipped, but that rule also counts upstream_failed state as 'done' which leads to the weird behaviour that  a task way down the dag will start running even though an upstream task failed and prevented all further tasks that it depends on to run.

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by felicity-cl** [comment]
all_ran trigger rule

**Comment by felicity-cl** [comment]
in here: https://github.com/apache/airflow/blob/main/airflow/ti_deps/deps/trigger_rule_dep.py#L72
add 
`
elif trigger_rule == TR.ALL_RAN:
            all_ran = (successes + failures) == done
            if not all_ran:
                yield self._failing_status(
                    reason=(
                        f"Task's trigger rule '{trigger_rule}' requires all upstream tasks to have run without skips or upstream failures "))
`

in here: https://github.com/apache/airflow/blob/69f6f9e01b6df76c3c8fa266d460324163957887/airflow/utils/trigger_rule.py#L1

add new rule ALL_RAN

**Comment by potiuk** [comment]
Converted that into a discussion. does not seem it caught up but feel free to provide a PR for that one to discuss. It should be easy to try.
